### PR TITLE
Add new ceph-fs configuration test

### PR DESCRIPTION
**Pull request by ethanmye-rs**
This PR adds functional tests for the charm-ceph-fs config options added in the following review:
https://review.opendev.org/c/openstack/charm-ceph-fs/+/842555

### CODE REVIEW

File: zaza/openstack/charm_tests/ceph/fs/tests.py [hunk 1]
```diff
@@ -103,6 +107,64 @@ def verify(stdin, stdout, stderr):
...
+        def _change_conf_check(mds_config):
+            """Change configs, then assert to ensure config was set.
+
+            Doesn't return a value.
+            """
+            loop = asyncio.get_event_loop()
+            crt = model.async_set_application_config('ceph-fs', mds_config)
+            loop.run_until_complete(crt)
+            results = _get_conf()
+            time.sleep(1)  # needs time to release lock and update conf.
```
Comment (lmlg) [inline zaza/openstack/charm_tests/ceph/fs/tests.py:142]: Using sleeps like this to assert values later can be rather fragile. I suggest using `tenacity` and retrying a couple of times to make the test more robust.

File: zaza/openstack/charm_tests/ceph/fs/tests.py [hunk 2]
```diff
@@ -103,6 +107,64 @@ def verify(stdin, stdout, stderr):
...
+            assert int(results['mds cache memory limit']) == \
+                   int(mds_config['mds-cache-memory-limit'])
+            assert float(results['mds cache reservation']) == \
```
Comment (lmlg) [inline zaza/openstack/charm_tests/ceph/fs/tests.py:145]: Comparing floats can be very error prone. Here I'd use `self.assertAlmostEqual`.

**Comment by ChrisMacNaughton** [comment]
This isn't really a functional test; it's purely a test that config was written to a file, without any verification that Ceph even cares about that config line. Broadly, the desired approach for functional testing is to verify that the things you're trying to do actually causes the desired effect; as a basic example: testing that a CephFS share can be created, written to, and read from is infinitely more valuable than testing that we wrote the config we think we did.

It's possible to get the same confidence int his change via unit-testing in the charm itself, rather than `cat`ing the ceph.conf to ensure that the expected changes appeared.

**Comment by ethanmye-rs** [comment]
> This isn't really a functional test; it's purely a test that config was written to a file, without any verification that Ceph even cares about that config line.

In my experience, Ceph ignores invalid config options in the .conf file, but that is anecdotal. From our earlier discussion, I agree querying the Ceph daemon directly is a better test to ensure Ceph actually understood and accepted the config values passed.

> It's possible to get the same confidence int his change via unit-testing in the charm itself, rather than `cat`ing the ceph.conf to ensure that the expected changes appeared.

I've left the unit tests as they are in the opendev repo and updated this repo with the 'functional' tests that ask Ceph for the configuration directly. Thanks for the help!

File: zaza/openstack/charm_tests/ceph/fs/tests.py [hunk 3]
```diff
@@ -103,6 +106,73 @@ def verify(stdin, stdout, stderr):
...
+            """
+            configs = ["mds_cache_memory_limit",
+                       "mds_cache_reservation",
+                       "mds_health_cache_threshold"]
+            holder = {}
+            for config in configs:
+                cmd = "sudo ceph daemon mds." \
+                      "$HOSTNAME config show | grep {}".format(config)
+                conf = model.run_on_unit(self.TESTED_UNIT, cmd)
+                print(conf)
```
Comment (ChrisMacNaughton) [inline zaza/openstack/charm_tests/ceph/fs/tests.py:127]: It's generally preferred to use a log statement, rather than `print`

**Review by ChrisMacNaughton — COMMENTED** [review]
I just have one issue with the tests now, it looks like a debug print was left.

File: zaza/openstack/charm_tests/ceph/fs/tests.py (hunk 3 above)
Comment (ethanmye-rs) [reply zaza/openstack/charm_tests/ceph/fs/tests.py:127]: Good catch -- I just removed it, not really needed outside debugging.

**Review by ChrisMacNaughton — APPROVED** [review]

**Pull request merged**